eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/RulesTests.hs

```diff
@@ -520,3 +542,50 @@ testRwdAliceSignsAlone''' =
         )
         (Coin 0)
         [asWitness Cast.alicePay, asWitness Cast.bobPay]
+
+-- | The reward aggregation bug described in the Shelley ledger spec in
+-- section 17.4 (in the Errata) resulted in needing to use 'aggregatedRewards' to change
+-- the behavior of how rewards are collected starting at protocol version 3.
+-- Instead of collecting a `Coin` for each stake credential, we collect 'Set Reward'.
+-- In major protocol version 2, it is impossible for this set to be empty, but sadly this
+-- property is not enforced in the types. For this reason, the property test
+-- 'propTickfPerservesLedgerView' removes these empty sets from an otherwise arbitrary
+-- 'NewEpochState'.
+filterEmptyRewards :: NewEpochState Shelley -> NewEpochState Shelley
+filterEmptyRewards (NewEpochState el bprev bcur es ru pd stash) =
+  NewEpochState el bprev bcur es ru' pd stash
+  where
+    removeEmptyRewards = Map.filter $ not . Set.null
+    ru' = case ru of
+      SNothing -> SNothing
+      SJust (Pulsing _ _) -> SNothing
+      SJust (Complete rewardUpdate) ->
+        SJust . Complete $ rewardUpdate {rs = removeEmptyRewards (rs rewardUpdate)}
+
+setDepositsToObligation :: NewEpochState Shelley -> NewEpochState Shelley
+setDepositsToObligation (NewEpochState el bprev bcur (EpochState as ss ls ppp pp nm) ru pd stash) =
+  NewEpochState el bprev bcur (EpochState as ss ls' ppp pp nm) ru pd stash
+  where
+    LedgerState (UTxOState utxo _deposited fees ppups stakeDistr) dpstate = ls
+    deposited = obligationDPState dpstate
+    ls' = LedgerState (UTxOState utxo deposited fees ppups stakeDistr) dpstate
```

Comment:
FYI, I am not suggesting to change this implementation, I just wanted to make a comment that is related to our discussion about pattern matches vs record syntax. Precisely in this kind of patterns it would be much clearer for the future reader and the current reader of the code to understand what is going on, because it is obvious what is being modified and which fields are needed for this update. All because we remove all the unnecessary noise. More over it  would make it future proof (i.e. account for field reordering, addition or removal):
```suggestion
setDepositsToObligation nes = nes { nesEs = es { esLState = ls { lsUTxOState =  utxoState }} }
  where
    es = nesEs nes
    ls = esLState es
    utxoState = (lsUTxOState ls) { utxosDeposited = obligationDPState $ lsDPState ls }
```

Naturally, with lenses this would be even more clear. Maybe in the future will define some lenses for all our types.